### PR TITLE
2016 day 19

### PR DESCRIPTION
**Pull request by Herrera93**
Here is the [problem description](https://adventofcode.com/2016/day/19) so everyone can understand the context.

Part 1's solution is based on [Josephus Problem](https://en.wikipedia.org/wiki/Josephus_problem), I found an interesting [video](https://www.youtube.com/watch?v=uCsD3ZGzMgE) about it, I encourage you to watch it so you can understand the solution.

In part 2's solution my first solution was to create:
```Rust
let mut v: Vec<usize> = (1..=n).collect();
```
where `n` is the number of elves and a for loop to calculate which index to delete based on v's length and current turn:
```Rust
let mut delete_index = turn + v.len() / 2;
if delete_index >= v.len() {
     delete_index -= v.len();
}
```
where `turn` is number in `0..v.len()`, this meant that I had to remove an element every iteration for `n-1` times. This was very inefficient once I tried it on the problem's input which I ran for 2 minutes and decided to stop it.

Now I tried to see if there was any kind of elegant solution like the Josephus Problem, so I started solving the problem by hand with different number of `n`.

In the problem's description they solve the problem when `n = 5`, this are the numbers that are safe in the first iteration `[1, 2, 4]`. Using `n = 8` the safe numbers in the first iteration `[1, 2, 3, 4, 7]`. After representing this on a table:

| n        | Safe numbers           |
--- | --- |
| `2` | `[1]` | 
| `3` | `[1, 3]`    | 
| `4` | `[1, 2]`     | 
| `5` | `[1, 2, 4]`     |  
| `6` | `[1, 2, 3, 6]`     | 
| `7` | `[1, 2, 3, 5]`     | 
| `8` | `[1, 2, 3, 4, 7]`     | 
| `9` | `[1, 2, 3, 4, 6, 9]`     | 
| `10` | `[1, 2,  3, 4, 5, 8]`     | 
| `11` | `[1, 2, 3, 4, 5, 7, 10]`     | 
| `12` | `[1, 2, 3, 4, 5, 6, 9]`     | 
| `13` | `[1, 2, 3, 4, 5, 6, 8, 11]`     | 
| `14` | `[1, 2, 3, 4, 5, 6, 7, 10, 13]`     | 
| `15` | `[1, 2, 3, 4, 5, 6, 7, 9, 12, 15]`     |
| `16` | `[1, 2, 3, 4, 5, 6, 7, 8, 11, 14]`     |
| `17` | `[1, 2, 3, 4, 5, 6, 7, 8, 10, 13, 16]`     |
| `18` | `[1, 2, 3, 4, 5, 6, 7, 8, 9, 12, 15, 18]`     |
| `19` | `[1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 14, 17]`     |
| `20` | `[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 13, 16, 19]`     |

We can start to see some patterns in the safe numbers, the first one is that every two `n`'s the number of starting safe numbers increases. In `4` and `5` the starting safe numbers are `[1, 2]` and in `6` and `7` are `[1, 2, 3]` and this goes on. 

For the next part there is also a pattern you see how the starting numbers of `5` are `[1, 2]` skips one number (`3`) to get the next safe number `4`, in `6` the starting numbers are `[1, 2, 3]` and then skip two numbers (`4` and `5`) to get the next safe number `6`. This alternates so even numbers skip two numbers and odd skip one number.

The last pattern is the easiest one but requires us to see the safe numbers of the higher `n`. When `n` is `20` we know that the starting safe numbers are `[1, 2, 3, 4, 5, 6, 8, 9, 10]`, since this is even we skip two numbers and add `13`. Now to determine the last few numbers we just skip two numbers until we reach the end. That means the safe numbers are `[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 13, 16, 19]`.

Now before executing this process again in the safe numbers we have to rearrange them because it's not `1`'s turn when we complete the first iteration. For example when `n` is `5` our safe numbers are `[1, 2, 4]` but if we apply the process again the result will be `[4]` which is not the correct answer.

If we see a table with the correct new starting indices we can also see a pattern there:

| n | New starting index |
--- | --- |
| `2` | `0` | 
| `3` | `2` | 
| `4` | `0` | 
| `5` | `3` | 
| `6` | `2` | 
| `7` | `4` | 
| `8` | `3` | 
| `9` | `3` | 
| `10` | `4` |
| `11` | `4` | 
| `12` | `4` | 
| `13` | `5` |  

We can see that after `n` is `8` the behavior is consistent (3 repeated number and then increase by 1), so what happens before `8`? Well first thing is that `6` is actually correct, as the rest of the number I can't explain, if anyone has an explanation feel free to post it. My theory is that since the consistent behavior resulting number for every number less than `8` (except for `6`) doesn't exists it ignores it.

Even if we don't know why this is happening we can still use this information to solve this problem.
First we'll need to find the starting safe numbers, for that we can use the following formula:

```Rust
let i = (v.len() + 2) / 3;
```

You may notice that this number doesn't include all the the starting safe numbers, for example when `n` is `9` this will return the first `i = (9 + 2) / 3 = 3` numbers `[1, 2, 3]` when we know that first starting safe numbers for `9` are `[1, 2, 3, 4]`, this is because this will help us find the new start for the next iteration since we know that the new start should be `4`. 

Let's also take into consideration when `n` is `18`, `i` would give use the first six safe numbers `[1, 2, 3, 4, 5, 6]` but there are three more sequential numbers `[7, 8, 9]`. For that we have this formula:

```Rust
let j = v.len() / 2;
```

The range `i..j` will give us the last sequential numbers. Now we can append this to our new vector for the next iteration since we know `i` is the new starting point. After that we need to find how many numbers will be skipped, we can use this:

```Rust
let inc = 2 - v.len() % 2;
```

The number `inc` will alternate between 1 or 2 depending if it's even or odd. We can add the next part to our new vector `v[j + inc..].step(3)`. This will skip the number(s) and will keep skipping two numbers. For the last part we can just append  the first starting safe number since we already processed those numbers `v[..i]`.

Now if we repeat this process on our resulting vector until the vector is of size `1`, we will find our solution. 

## PR - Merge Checklist
- [ ] CR'd
- [ ] Workflow: Trello card is in "Code Review" column
- [ ] README updated or N/A
- [ ] Benchmarks run and compared with master or N/A

### CODE REVIEW

**Comment by Herrera93** [comment]
[2016-day-19-part-1](https://trello.com/c/gBuc0soz/124-2016-day-19-part-1)

**Comment by Herrera93** [comment]
[2016-day-19-part-2](https://trello.com/c/vppYHzE8/149-2016-day-19-part-2)

**Comment by lifehug** [comment]
This is a fantastic write up, just wanted to mention I appreciate the effort you put into making it clear for us.

File: src/year2016/day_19.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+use itertools::Itertools;
+
+pub fn who_wins(elves: usize) -> usize {
+    // This is a solution based on Josephus Problem
```
Comment (lifehug) [inline src/year2016/day_19.rs]: wow, what a concise and simple solution

Comment (Herrera93) [reply src/year2016/day_19.rs]: I know, the binary solution blew my mind.

**Pull request merged**